core/rate-limit.go

```diff
@@ -0,0 +1,221 @@
+	if q.front <= q.back {
+		return q.back - q.front
+	} else {
+		return q.back - q.front + q.size + 1
+	}
+}
+
+func (q queue) Head() int64 {
+	return q.data[q.front]
+}
+
+func (q queue) Tail() int64 {
+	if q.back == 0 {
+		return q.data[q.size]
+	} else {
+		return q.data[q.back-1]
+	}
+}
+
+func (q queue) Full() bool {
+	return q.Length() >= q.size
+}
+
+func (q *queue) Push(x int64) {
+	if q.Full() {
+		// Fails silently; caller must check
+		return
+	}
+
+	q.data[q.back] = x
+	q.back = (q.back + 1) % (q.size + 1)
+}
+
+func (q *queue) Pop() (x int64) {
+	if q.Length() == 0 {
+		// Fails silently; caller must check
+		return
+	}
+
+	x = q.data[q.front]
+	q.front = (q.front + 1) % (q.size + 1)
+	return
+}
+
+func (q *queue) Clear() {
+	q.front = 0
+	q.back = 0
+}
+
+// A RateLimiter limits each identified thing to a certain
+// number of attempts per ID per unit time.  It is defined
+// by the following parameters:
+// * numQueues - the number of queues it will maintain
+// * queueSize - the number of attempts remembered for each ID
+// * Window - how long each entry remains in the queue
+//
+// In effect, the each of `numQueues` IDs is limited to no
+// more than queueSize attempts within the Window.
+//
+// Each time an attempt is made for an ID:
+// * If the ID has a queue and there is space in it, the attempt
+//   is logged in the queue for that ID
+// * If there are more queues remaining, a new queue is assigned
+//   for that ID, and the attempt is logged there
+// * Otherwise, the attempt is denied and forgotten
+//
+// On every attempt, we clean out old entries so that queue space
+// and queues are freed up. All memory is pre-allocated on
+// construction, so the size of this structure is constant.
+//
+// To use this structure, you may need to define a map from your
+// identifier of choice to an integer ID value
+//
+// Application-layer things that might make sense as rate-
+// limiting identifiers:
+// * Account keys (at new-reg)
+// * Email addresses (at new-reg)
+// * label.PSL names (at new-authz)
+// * full names (at new-cert, on a different time scale than the above)
+type RateLimit struct {
+	numQueues int
+	queueSize int
+	Window    time.Duration
+
+	labels map[int64]int // id -> queue index
+	used   map[int]int64 // queue index -> id
+	free   queue
+	queues []queue
+}
+
+func NewRateLimit(numQueues, queueSize int, window time.Duration) RateLimit {
+	rl := RateLimit{}
+	rl.Resize(numQueues, queueSize, window)
+	return rl
+}
+
+func (rl *RateLimit) Resize(numQueues, queueSize int, window time.Duration) {
+	// Reset all the state variables
+	rl.numQueues = numQueues
+	rl.queueSize = queueSize
+	rl.Window = window
+
+	rl.labels = map[int64]int{}
+	rl.used = map[int]int64{}
+
+	rl.free = newQueue(numQueues)
+	rl.queues = make([]queue, numQueues, numQueues)
+	for i := 0; i < numQueues; i += 1 {
+		rl.queues[i] = newQueue(queueSize)
+		rl.free.Push(int64(i))
+	}
+	return
+}
+
+func (rl *RateLimit) Trim() {
```

Comment:
This method is going to need to be synchronized; it's going to be the most obvious offender for deleting things out from under other HTTP connections.